message_endpoint.go

```diff
@@ -0,0 +1,225 @@
+package swim
+
+import (
+	"errors"
+
+	"github.com/DE-labtory/swim/pb"
+	"github.com/gogo/protobuf/proto"
+	"github.com/it-chain/iLogger"
+
+	"time"
+)
+
+var ErrSendTimeout = errors.New("Error send timeout")
+var ErrUnreachable = errors.New("Error this shouldn't reach")
+var ErrInvalidMessage = errors.New("Error invalid message")
+
+// Callback is called when target member sent back to local member a message
+type callback func(msg pb.Message)
+
+// RespondHandler manages callback functions
+type respondHandler struct {
+	callbacks map[string]callback
+}
+
+func NewRespondHandler() *respondHandler {
+	return &respondHandler{
+		callbacks: make(map[string]callback),
+	}
+}
+
+func (r *respondHandler) addCallback(seq string, cb callback) {
+	r.callbacks[seq] = cb
+}
+
+// Handle, each time other member sent back
+// a message, callback matching message's seq is called
+func (r *respondHandler) handle(msg pb.Message) {
+	seq := msg.Seq
+	cbFn := r.callbacks[seq]
+
+	if cbFn == nil {
+		iLogger.Panicf(nil, "")
+	}
+
+	cbFn(msg)
+	delete(r.callbacks, seq)
+}
+
+func (r *respondHandler) hasCallback(seq string) bool {
+	for s := range r.callbacks {
+		if seq == s {
+			return true
+		}
+	}
+	return false
+}
+
+type MessageEndpointConfig struct {
+	EncryptionEnabled  bool
+	DefaultSendTimeout time.Duration
+}
+
+// MessageEndpoint basically do receiving packet and determine
+// which logic executed based on the packet.
+type MessageEndpoint struct {
+	config         MessageEndpointConfig
+	transport      UDPTransport
+	messageHandler MessageHandler
+	awareness      Awareness
+	resHandler     *respondHandler
```

Comment:
resHandler is reference but Why messageHandler is not reference?!

Reply:
I coded like this because I think that each `MessageEndpoint`should reflect respondHandler's change. and I've remembered interface as pointer of struct, and MessageHandler is interface, resHandler struct